### PR TITLE
add nim language parser

### PR DESCRIPTION
**Pull request by jangko**
I use it daily, I hope others can get benefit too from this parser.

### CODE REVIEW

**Comment by vhda** [comment]
Please fix indentation of modified files.
Note: did not review new file `nim.c`.

**Comment by vhda** [comment]
I know I'm being picky :) but could you squash the fix over the original commit and `git push -f`?

**Comment by jangko** [comment]
i am not good at pushing things around, especially in github land. can you give me a hand?
or please tell me how to do that :) (i'm still figuring it out)
thanks

**Comment by steveno** [comment]
@jangko 
You can check out this page for a good example of how to squash commits
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Once you've squashed all your commits, simply do a `git push -f` to your fork, and github will update the pull request for you automatically.

**Comment by jangko** [comment]
I think i managed to squash them, thanks @steveno, @vhda

**Comment by vhda** [comment]
You might want to use `verbose()` instead of commenting out the debug printing messages. IMHO that would make the code prettier and the message display could be easily be enabled in run time.

**Comment by ffes** [comment]
Isn't `verbose()` just for the `--verbose` command line switch? Look at debug.c for various helper functions.

Please don't leave code in the final version that is commented out and just there for testing and trying things.

**Comment by vhda** [comment]
@ffes thanks for correcting me!

**Comment by jangko** [comment]
OK, using debug(...) from debug.h and -D2 cmdline switch to activate it, 
fix some debug mode in **mk_mingw.mak** and **mk_mvc.mak**
cleanup unnecessary comments, squash all previous commits

**Comment by ffes** [comment]
> fix some debug mode in **mk_mingw.mak** and **mk_mvc.mak**

Great! I never used these debug builds on Windows. When I need to debug, I always use the debugger of Visual Studio.

File: mk_mingw.mak [hunk 1]
```diff
@@ -16,8 +16,8 @@ VPATH = . ./main ./parsers
 ctags.exe: OPT = -O4 -Os -fexpensive-optimizations
 ctags.exe: LDFLAGS = -s
 dctags.exe: OPT = -g
-dctags.exe: DEBUG = -DDEBUG
-dctags.exe: SOURCES += debug.c
+dctags.exe: CFLAGS += -DDEBUG
+# dctags.exe: SOURCES += $(MAIN_DIR)/debug.c
```
Comment (ffes) [inline mk_mingw.mak]: Could you remove the `#`?

I would say the default behavior of a debug build should be showing debug messages. If you want to disable it yourself and still use the debug build, do so in your personal branch.

File: main/vstring.h [hunk 2]
```diff
@@ -22,7 +22,7 @@
 *   MACROS
 */
 #ifndef DEBUG
-# define VSTRING_PUTC_MACRO 1
+#define VSTRING_PUTC_MACRO 1
```
Comment (ffes) [inline main/vstring.h]: Everywhere in the code base there are these spaces to indent. See `debug.h` or `general.h` as examples. So please that kind of thing as they are.

File: mk_mingw.mak (hunk 1 above)
Comment (jangko) [inline mk_mingw.mak]: in source.mak line 106, debug.c already included, 
so it is unecessary to put it in mk_mingw.mak (the compiler will complaint anyway)

DEBUG_SOURCES = $(MAIN_DIR)/debug.c

SOURCES = $(MAIN_SOURCES) $(PARSER_SOURCES) $(DEBUG_SOURCES)

Comment (ffes) [inline mk_mingw.mak]: Ah, missed that. Thanks for the clarification.

In that case there is no need to keep the line.

**Comment by masatake** [comment]
@jangko, thank you very much. 

valgrind is avaiable on your environment?

```
make units LANGUAGES=Nim VG=1
./config.status --header config.h
config.status: creating config.h
config.status: config.h is unchanged

Category: ROOT
------------------------------------------------------------
Testing nim-test_input1 as Nim                              passed
Testing nim-test_input2 as Nim                              failed (valgrind-error)
```

**Comment by masatake** [comment]
```
diff --git a/parsers/nim.c b/parsers/nim.c
index fd9c3f9..a2d2b70 100644
--- a/parsers/nim.c
+++ b/parsers/nim.c
@@ -1781,6 +1781,7 @@ static int considerStrongSpaces(nimToken* tok, boolean strongSpaces, int x)
 static int considerAsgn(nimToken* tok, int value)
 {
   int L = vStringLength(tok->literal);
+  Assert(L > 0);
   return (tok->literal->buffer[L-1] == '=' ? 1 : value);
 }
 
```

**Comment by jangko** [comment]
@masatake : I never use valgrind before coz I work on windows machine.
but I'll try to setup a linux one.

**Comment by masatake** [comment]
nim.c is the guest *.c file. Which is larger than the c.c.
By mering this branch with enable-gcov branch, I checked the coverage of units:
```
File 'parsers/nim.c'
Lines executed:52.35% of 1721
Creating 'nim.c.gcov'
```

**Comment by masatake** [comment]
@jangko, see the assertion I inserted. It seems that L is 0.

**Comment by masatake** [comment]
I'm soory I mistakenly merger your PR though more revising are needed.
Could you open a new PR?

**Comment by masatake** [comment]
(Can I open this PR? I cannot find the way.)

**Comment by jangko** [comment]
of course i can open a new PR.

**Comment by masatake** [comment]
(Thank you. It seems that I am tired...)

**Pull request merged**